core/src/main/scala/scala/macros/package.scala

```diff
@@ -1,21 +1,194 @@
 package scala
 
-package object macros
-    extends scala.macros.config.Api
-    with scala.macros.config.Aliases
-    with scala.macros.inputs.Api
-    with scala.macros.inputs.Aliases
-    with scala.macros.prettyprinters.Api
-    with scala.macros.prettyprinters.Aliases
-    with scala.macros.trees.Api
-    with scala.macros.trees.Aliases
-    with scala.macros.semantic.Api
-    with scala.macros.semantic.Aliases
-    with scala.macros.Universe {
+package object macros {
 
-  private[macros] val universe = new ThreadLocal[scala.macros.Universe]
-  private[macros] def abstracts = {
-    if (universe.get == null) sys.error("this API can only be called in a macro expansion")
-    universe.get.abstracts.asInstanceOf[Abstracts]
+  private[macros] val universeStore = new ThreadLocal[core.Universe]
+  private[macros] def universe: ghost.type = {
+    val result = universeStore.get
+    if (result == null) sys.error("this API can only be called in a macro expansion")
+    else result.asInstanceOf[ghost.type]
+  }
+
+  // Note(fengyun): required for Dotty, otherwise `!t` will infer Nothing for `B`,
+  // causing run-time cast exceptions
+  private[macros] val ghost: core.Universe = null
+
+  private implicit class XtensionBang[A](val a: A) extends AnyVal {
+    @inline def unary_![B]: B = a.asInstanceOf[B]
+  }
```

Comment:
It seems remove `@inline` is less hacky than introducing `ghost`. WDYT?

Reply:
Agreed, removed inline.